File: tests/test_profile_tone_compose.py
"""The NVDA driver's voicing tone for a voice profile (3.10 beta 10, #124).

Runs the real VoicingToneMixin._applyVoicingTone from the add-on with stand-ins
for NVDA, the frontend and the player, and checks the contract:

- a profile with its own voicingTone block: every stored value reaches the
  player with the listener's settings at neutral; a moved setting composes
  with it; moving it back restores it;
- a built-in voice, or a profile without a block: exactly the values the
  driver produced before the change (the settings are the values);
- a David-derived profile keeps its stored head size at the neutral setting
  (the built-in David's head-size 100 lives inside the profile now).

No NVDA, DLL or audio is needed.
"""
from __future__ import annotations

import math
import pathlib
import types

import pytest

REPO = pathlib.Path(__file__).resolve().parent.parent
SRC = REPO / "nvdaAddon" / "synthDrivers" / "tgSpeechBox" / "voicing_tone.py"

TONE_FIELDS = (
    "voicingPeakPos", "voicedPreEmphA", "voicedPreEmphMix",
    "highShelfGainDb", "highShelfFcHz", "highShelfQ",
    "voicedTiltDbPerOct", "noiseGlottalModDepth",
    "pitchSyncF1DeltaHz", "pitchSyncB1DeltaHz",
    "speedQuotient", "aspirationTiltDbPerOct", "cascadeBwScale", "tremorDepth",
    "nasalBwScale", "f4FreqScale", "nasalGainScale",
    "chorusDepth", "chorusDetuneHz",
)
DSP_DEFAULTS = dict(
    voicingPeakPos=0.91, voicedPreEmphA=0.92, voicedPreEmphMix=0.35,
    highShelfGainDb=4.0, highShelfFcHz=2000.0, highShelfQ=0.7,
    voicedTiltDbPerOct=0.0, noiseGlottalModDepth=0.0,
    pitchSyncF1DeltaHz=0.0, pitchSyncB1DeltaHz=0.0,
    speedQuotient=2.0, aspirationTiltDbPerOct=0.0, cascadeBwScale=0.9, tremorDepth=0.0,
    nasalBwScale=1.0, f4FreqScale=1.0, nasalGainScale=1.0,
    chorusDepth=0.0, chorusDetuneHz=2.0,
)


class _Tone:
    def __init__(self, **kw):
        for k in TONE_FIELDS:
            setattr(self, k, kw.get(k, DSP_DEFAULTS[k]))

    @classmethod
    def defaults(cls):
        return cls()


def _load_mixin():
    src = SRC.read_text(encoding="utf-8")
    src = src.replace("from logHandler import log", "")
    src = src.replace("from . import speechPlayer", "")
    src = src.replace("from .constants import VOICE_PROFILE_PREFIX", "VOICE_PROFILE_PREFIX = 'profile:'")
    ns = {
        "__name__": "voicing_tone_under_test",
        "log": types.SimpleNamespace(error=lambda *a, **k: None, debug=lambda *a, **k: None,
                                     warning=lambda *a, **k: None, info=lambda *a, **k: None),
        "speechPlayer": types.SimpleNamespace(VoicingTone=_Tone),
    }
    exec(compile(src, str(SRC), "exec"), ns)
    return ns["VoicingToneMixin"]


VoicingToneMixin = _load_mixin()


class _Player:
    def __init__(self):
        self.tone = None

    def hasVoicingToneSupport(self):
        return True

    def setVoicingTone(self, tone):
        self.tone = tone


class _Frontend:
    def __init__(self, profile_tone):
        self.profile_tone = profile_tone  # None = no voicingTone block

    def setVoiceProfile(self, name):
        pass

    def hasExplicitVoicingTone(self):
        return self.profile_tone is not None

    def getVoicingTone(self):
        return self.profile_tone


# Slider neutral positions in the driver.
NEUTRAL = dict(_curVoiceTilt=50, _curNoiseGlottalMod=0, _curPitchSyncF1=50, _curPitchSyncB1=50,
               _curSpeedQuotient=50, _curAspirationTilt=50, _curCascadeBwScale=50,
               _curVoiceTremor=0, _curHeadSize=50, _curChorusDepth=0, _curChorusDetune=33)


def _run(profile_tone, profile_name="Probe", **sliders):
    class Driver(VoicingToneMixin):
        def _pushFrameExDefaultsToFrontend(self):
            pass

    d = Driver()
    d._player = _Player()
    d._frontend = _Frontend(profile_tone)
    for k, v in {**NEUTRAL, **sliders}.items():
        setattr(d, k, v)
    d._applyVoicingTone(profile_name)
    assert d._player.tone is not None
    return d._player.tone


def _frontend_tone(**stored):
    """What nvspFrontend_getVoicingTone returns: its defaults plus the stored keys."""
    base = dict(DSP_DEFAULTS, highShelfGainDb=5.5, cascadeBwScale=1.0)
    base.update(stored)
    return _Tone(**base)


# Every composable key non-default (Astra's fixture values plus the rest).
STORED = dict(voicingPeakPos=0.88, voicedPreEmphA=0.9, voicedPreEmphMix=0.22,
              highShelfGainDb=2.0, highShelfFcHz=2400.0, highShelfQ=0.8,
              voicedTiltDbPerOct=-10.0, noiseGlottalModDepth=0.3,
              pitchSyncF1DeltaHz=12.0, pitchSyncB1DeltaHz=8.0,
              speedQuotient=1.6, aspirationTiltDbPerOct=3.0, cascadeBwScale=0.8, tremorDepth=0.1,
              nasalBwScale=1.3, f4FreqScale=1.08, nasalGainScale=1.1)


def _old_builtin(**s):
    """The driver's values for a built-in voice before the change."""
    s = {**NEUTRAL, **s}
    sq = s["_curSpeedQuotient"]
    sq = 0.5 + (sq / 50.0) * 1.5 if sq <= 50 else 2.0 + ((sq - 50.0) / 50.0) * 2.0
    bw = s["_curCascadeBwScale"]
    bw = 2.0 - (bw / 50.0) if bw <= 50 else 1.0 - ((bw - 50.0) / 50.0) * 0.7
    hs = s["_curHeadSize"]
    f4 = 1.25 - (hs / 50.0) * 0.25 if hs <= 50 else 1.0 - ((hs - 50.0) / 50.0) * 0.15
    return dict(
        voicedTiltDbPerOct=max(-24.0, min(24.0, (s["_curVoiceTilt"] - 50.0) * 0.48)),
        noiseGlottalModDepth=s["_curNoiseGlottalMod"] / 100.0,
        pitchSyncF1DeltaHz=(s["_curPitchSyncF1"] - 50.0) * 1.2,
        pitchSyncB1DeltaHz=(s["_curPitchSyncB1"] - 50.0) * 1.0,
        speedQuotient=sq,
        aspirationTiltDbPerOct=(s["_curAspirationTilt"] - 50.0) * 0.24,
        cascadeBwScale=max(0.3, min(2.0, bw)),
        tremorDepth=max(0.0, min(0.5, s["_curVoiceTremor"] / 100.0 * 0.4)),
        f4FreqScale=max(0.7, min(1.5, f4)),
        chorusDepth=s["_curChorusDepth"] / 100.0,
        chorusDetuneHz=0.5 + s["_curChorusDetune"] / 100.0 * 4.5,
    )


def test_profile_values_reach_the_player_at_neutral_settings():
    tone = _run(_frontend_tone(**STORED))
    for k, v in STORED.items():
        assert math.isclose(getattr(tone, k), v, abs_tol=1e-12), k


@pytest.mark.parametrize("sliders,field,expected", [
    (dict(_curSpeedQuotient=25), "speedQuotient", 1.6 * (1.25 / 2.0)),
    (dict(_curHeadSize=75), "f4FreqScale", 1.08 * 0.925),
    (dict(_curVoiceTilt=60), "voicedTiltDbPerOct", -10.0 + 4.8),
    (dict(_curNoiseGlottalMod=20), "noiseGlottalModDepth", 0.3 + 0.2),
    (dict(_curPitchSyncF1=60), "pitchSyncF1DeltaHz", 12.0 + 12.0),
    (dict(_curAspirationTilt=25), "aspirationTiltDbPerOct", 3.0 - 6.0),
    (dict(_curCascadeBwScale=75), "cascadeBwScale", 0.8 * 0.65),
    (dict(_curVoiceTremor=50), "tremorDepth", 0.1 + 0.2),
])
def test_a_moved_setting_composes_with_the_profile(sliders, field, expected):
    tone = _run(_frontend_tone(**STORED), **sliders)
    assert math.isclose(getattr(tone, field), expected, abs_tol=1e-9), field
    # The rest of the profile is untouched by that one setting.
    for k, v in STORED.items():
        if k != field:
            assert math.isclose(getattr(tone, k), v, abs_tol=1e-12), k


def test_returning_a_setting_to_neutral_restores_the_profile():
    _run(_frontend_tone(**STORED), _curSpeedQuotient=10, _curHeadSize=90)
    tone = _run(_frontend_tone(**STORED))
    assert math.isclose(tone.speedQuotient, 1.6) and math.isclose(tone.f4FreqScale, 1.08)


@pytest.mark.parametrize("sliders", [
    {}, dict(_curSpeedQuotient=30, _curHeadSize=80), dict(_curVoiceTilt=20, _curCascadeBwScale=90),
    dict(_curNoiseGlottalMod=40, _curPitchSyncF1=70, _curPitchSyncB1=10, _curVoiceTremor=30),
    dict(_curChorusDepth=50, _curChorusDetune=80),
])
@pytest.mark.parametrize("profile_name,profile_tone", [("", None), ("NoBlock", None)])
def test_built_in_voices_and_blockless_profiles_are_unchanged(sliders, profile_name, profile_tone):
    tone = _run(profile_tone, profile_name=profile_name, **sliders)
    for k, v in _old_builtin(**sliders).items():
        assert math.isclose(getattr(tone, k), v, abs_tol=1e-12), k
    for k in ("voicingPeakPos", "voicedPreEmphA", "voicedPreEmphMix", "highShelfGainDb",
              "highShelfFcHz", "highShelfQ", "nasalBwScale", "nasalGainScale"):
        assert math.isclose(getattr(tone, k), DSP_DEFAULTS[k], abs_tol=1e-12), k


def test_david_derived_profile_keeps_its_head_size():
    # The editor saves the built-in David's head size (slider 100) as f4FreqScale 0.85.
    tone = _run(_frontend_tone(f4FreqScale=0.85, voicedTiltDbPerOct=-3.0, highShelfGainDb=4.0),
                profile_name="David variant")
    assert math.isclose(tone.f4FreqScale, 0.85, abs_tol=1e-12)
    assert math.isclose(tone.voicedTiltDbPerOct, -3.0, abs_tol=1e-12)


def test_chorus_stays_a_listener_setting():
    tone = _run(_frontend_tone(**STORED), _curChorusDepth=40, _curChorusDetune=60)
    assert math.isclose(tone.chorusDepth, 0.4) and math.isclose(tone.chorusDetuneHz, 0.5 + 0.6 * 4.5)

File: tests/cpp/test_profile_tone_compose.cpp
// A voice profile's voice source through the shared composition rule
// (src/voicingToneCompose.h, used by SAPI, Android and iOS; the NVDA driver
// mirrors it and has its own pytest).  With the real frontend and packs:
// Beth's stored voicingTone, composed with neutral listener settings, comes
// out exactly as stored; a moved setting follows the rule; the frontend's
// fallbacks fill keys a profile leaves out.
#include <cmath>
#include <string>

#include "doctest.h"
#include "pack_fixture.h"
#include "nvspFrontend.h"
#include "voicingToneCompose.h"

namespace {

struct EnUsHandle : tgsb_test::HandleFixture {
    EnUsHandle() : tgsb_test::HandleFixture("en-us") {}
};

speechPlayer_voicingTone_t profileBase(const nvspFrontend_VoicingTone& pt) {
    speechPlayer_voicingTone_t t = speechPlayer_getDefaultVoicingTone();
    t.voicingPeakPos = pt.voicingPeakPos;
    t.voicedPreEmphA = pt.voicedPreEmphA;
    t.voicedPreEmphMix = pt.voicedPreEmphMix;
    t.highShelfGainDb = pt.highShelfGainDb;
    t.highShelfFcHz = pt.highShelfFcHz;
    t.highShelfQ = pt.highShelfQ;
    t.voicedTiltDbPerOct = pt.voicedTiltDbPerOct;
    t.noiseGlottalModDepth = pt.noiseGlottalModDepth;
    t.pitchSyncF1DeltaHz = pt.pitchSyncF1DeltaHz;
    t.pitchSyncB1DeltaHz = pt.pitchSyncB1DeltaHz;
    t.speedQuotient = pt.speedQuotient;
    t.aspirationTiltDbPerOct = pt.aspirationTiltDbPerOct;
    t.cascadeBwScale = pt.cascadeBwScale;
    t.tremorDepth = pt.tremorDepth;
    t.nasalBwScale = pt.nasalBwScale;
    t.f4FreqScale = pt.f4FreqScale;
    t.nasalGainScale = pt.nasalGainScale;
    return t;
}

void composeNeutral(speechPlayer_voicingTone_t& t) {
    speechPlayer_composeListenerSettings(&t, 0.0, 0.0, 0.0, 0.0, 2.0, 0.0, 1.0, 0.0, 1.0, 1.0, 1.0);
}

}  // namespace

TEST_CASE_FIXTURE(EnUsHandle, "profile tone: Beth's stored voice source survives neutral listener settings") {
    REQUIRE(nvspFrontend_setVoiceProfile(handle, "Beth"));
    nvspFrontend_VoicingTone pt{};
    REQUIRE(nvspFrontend_getVoicingTone(handle, &pt) == 1);
    // packs/phonemes.yaml, Beth's voicingTone
    CHECK(pt.voicedTiltDbPerOct == doctest::Approx(-10.0));
    CHECK(pt.speedQuotient == doctest::Approx(1.60));
    CHECK(pt.f4FreqScale == doctest::Approx(1.08));
    CHECK(pt.nasalBwScale == doctest::Approx(1.30));
    CHECK(pt.nasalGainScale == doctest::Approx(1.10));
    CHECK(pt.highShelfGainDb == doctest::Approx(5.5));  // left out -> the frontend's block fallback

    speechPlayer_voicingTone_t t = profileBase(pt);
    composeNeutral(t);
    CHECK(t.voicedTiltDbPerOct == doctest::Approx(-10.0));
    CHECK(t.speedQuotient == doctest::Approx(1.60));
    CHECK(t.f4FreqScale == doctest::Approx(1.08));
    CHECK(t.nasalBwScale == doctest::Approx(1.30));
    CHECK(t.nasalGainScale == doctest::Approx(1.10));
    CHECK(t.cascadeBwScale == doctest::Approx(1.0));
    CHECK(t.highShelfGainDb == doctest::Approx(5.5));
}

TEST_CASE_FIXTURE(EnUsHandle, "profile tone: Bobby's head size and speed quotient survive neutral settings") {
    REQUIRE(nvspFrontend_setVoiceProfile(handle, "Bobby"));
    nvspFrontend_VoicingTone pt{};
    REQUIRE(nvspFrontend_getVoicingTone(handle, &pt) == 1);
    speechPlayer_voicingTone_t t = profileBase(pt);
    composeNeutral(t);
    CHECK(t.f4FreqScale == doctest::Approx(1.20));
    CHECK(t.speedQuotient == doctest::Approx(2.20));
    CHECK(t.highShelfGainDb == doctest::Approx(2.0));
    CHECK(t.voicingPeakPos == doctest::Approx(0.88));
}

TEST_CASE("profile tone: moved settings compose, neutral restores") {
    speechPlayer_voicingTone_t base = speechPlayer_getDefaultVoicingTone();
    base.voicedTiltDbPerOct = -10.0;
    base.speedQuotient = 1.6;
    base.f4FreqScale = 1.08;
    base.noiseGlottalModDepth = 0.3;
    base.cascadeBwScale = 0.8;

    speechPlayer_voicingTone_t t = base;
    // tilt +4.8, speed quotient slider 25 (1.25), head size 75 (0.925), noise +0.2, sharpness 75 (0.65)
    speechPlayer_composeListenerSettings(&t, 4.8, 0.2, 0.0, 0.0, 1.25, 0.0, 0.65, 0.0, 1.0, 0.925, 1.0);
    CHECK(t.voicedTiltDbPerOct == doctest::Approx(-5.2));
    CHECK(t.speedQuotient == doctest::Approx(1.0));
    CHECK(t.f4FreqScale == doctest::Approx(0.999));
    CHECK(t.noiseGlottalModDepth == doctest::Approx(0.5));
    CHECK(t.cascadeBwScale == doctest::Approx(0.52));

    speechPlayer_voicingTone_t back = base;
    composeNeutral(back);
    CHECK(back.speedQuotient == doctest::Approx(1.6));
    CHECK(back.f4FreqScale == doctest::Approx(1.08));
    CHECK(back.voicedTiltDbPerOct == doctest::Approx(-10.0));

    // Clamps hold at the ends.
    speechPlayer_voicingTone_t hi = base;
    speechPlayer_composeListenerSettings(&hi, 40.0, 2.0, 0, 0, 4.0, 30.0, 2.0, 1.0, 5.0, 2.0, 5.0);  // -10 + 40 -> clamp 24
    CHECK(hi.voicedTiltDbPerOct == doctest::Approx(24.0));
    CHECK(hi.noiseGlottalModDepth == doctest::Approx(1.0));
    CHECK(hi.f4FreqScale == doctest::Approx(1.5));
    CHECK(hi.tremorDepth == doctest::Approx(0.5));
}

TEST_CASE_FIXTURE(EnUsHandle, "profile tone: a profile without a voicingTone block reports none") {
    // No profile: the hosts keep their built-in mapping.
    REQUIRE(nvspFrontend_setVoiceProfile(handle, ""));
    nvspFrontend_VoicingTone pt{};
    CHECK(nvspFrontend_getVoicingTone(handle, &pt) == 0);
}
